### PR TITLE
User code runner draft

### PR DESCRIPTION
**Pull request by weswigham**
Adds a new test runner which runs real projects with `npm i` and `tsc`, along with a test of this style for `tslint`. Run with `gulp runtests-parallel --test=user` (or `jake` equivalent)

@sandersn Has a bunch of tests to add to this.

### CODE REVIEW

**Comment by sandersn** [comment]
I think RWC or real-world is a misleading term for this, since we're not testing the projects themselves, but their consumers. What we really care about is that new versions of typescript don't break the projects' interfaces for consumers. Which is why, for example, tsconfig specifies `"strict": true`, because that's what we expect a lot of consumers to use.

File: tests/cases/realworld/tslint/package.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "author": "",
+  "license": "Apache-2.0",
+  "dependencies": {
+    "tslint": "^5.8.0",
+    "typescript": "^2.5.3"
+  },
+  "devDependencies": {
```
Comment (sandersn) [inline tests/cases/realworld/tslint/package.json]: should we also generate package.json in order to be able to copy `devDependencies` from the newly installed `tslint`? (Or just change these to latest, as you suggested in person :)

Comment (weswigham) [reply tests/cases/realworld/tslint/package.json]: Changed to `latest`.

**Comment by weswigham** [comment]
> I think RWC or real-world is a misleading term for this, since we're not testing the projects themselves, but their consumers.

Have a decent name? I'll take anything; this was pretty much just supposed to be a public version of the RWC suite without all the capturing complexity done for RWC.

**Comment by sandersn** [comment]
* breaking
* consumer
* example
* project (though I think this is already taken)
* integration (though I feel like this is not quite right)

any of the above + "_api"

**Comment by weswigham** [comment]
@sandersn @DanielRosenwasser said "Public RWC" (making the shortname used internally "publicrwc")

**Comment by sandersn** [comment]
meh. I'm not a big fan of the name because it's (1) not (currently) the same infrastructure as the RWC (2) the RWC is over half public code now anyway (3) not (currently) the same intent as the RWC.

(4) I'm going to pronounce it "publikwerk" whether I want to or not.

But  besides that, I don't care much.

**Comment by weswigham** [comment]
> (1) not (currently) the same infrastructure as the RWC (2) the RWC is over half public code now anyway (3) not (currently) the same intent as the RWC.

It's technically different, yes, but the intent is still to test code... from... the real... world.......

What about `"users"`? Mostly because `tests/cases/users/<tslint or whatever>` reads pretty okay, and `UsersRunner` or `UserCodeRunner` doesn't sound like the _worst_ name.

**Comment by sandersn** [comment]
I like users fine, although "user" is even better. I guess my naming argument rests mostly on "real-world code" already being taken by the RWC, which is not that great an argument.

**Comment by weswigham** [comment]
Alright, `user` it is.

**Comment by sandersn** [comment]
@DanielRosenwasser now that this is ready, you might want to take a look. Then let's have a conversation about what else to add, plus how complicated to make the tests, especially a framework like vue.

**Comment by weswigham** [comment]
> Then let's have a conversation about what else to add, plus how complicated to make the tests, especially a framework like vue.

It's also worth noting that we can maintain complex tests in an external repo an keep a bit submodule reference to it here.

**Pull request merged**